src/openforms/submissions/tasks/__init__.py

```diff
@@ -0,0 +1,81 @@
+from celery import chain
+from celery.result import AsyncResult
+
+from openforms.celery import app
+
+from ..models import Submission
+from .appointments import *  # noqa
+from .emails import *  # noqa
+from .pdf import *  # noqa
+from .registration import *  # noqa
+from .user_uploads import *  # noqa
+
+
+def on_completion(submission_id: int) -> str:
+    """
+    Celery chain of tasks to execute on a submission completion event.
+
+    This SHOULD be invoked as a transaction.on_commit(...) handler, therefore it should
+    not execute any extra queries in the process this function is running in.
+    """
+    # use immutable signatures so that the result of previous tasks is not passed
+    # in as an argument to chained tasks
+    register_appointment_task = maybe_register_appointment.si(submission_id)
+    update_appointment_task = maybe_update_appointment.si(submission_id)
+    generate_report_task = generate_submission_report.si(submission_id)
+    register_submission_task = register_submission.si(submission_id)
+    obtain_submission_reference_task = obtain_submission_reference.si(submission_id)
+    finalize_completion_task = finalize_completion.si(submission_id)
+
+    # for the orchestration with distributed processing and dependencies between
+    # tasks, see the Celery documentation:
+    # https://docs.celeryproject.org/en/stable/userguide/canvas.html#guide-canvas
+    # Additionally, check the developer documentation of Open Forms for a visual
+    # representation of the task orchestration and dependencies (TODO, @joeri)
+    # The linked task (= next task) is only executed if the previous task returns
+    # successfully, so error handling needs to happen inside each task.
+    on_completion_chain = chain(
+        # The appointment must be registered before a confirmation PDF is generated and
+        # any backend registration happens, as on-failure, the user should get feedback
+        # about the failure.
+        register_appointment_task,
+        # The submission report needs to already have been generated before it can be
+        # attached in the registration backend.
+        # TODO: can be in parallel with register_appointment_task and if that fails -> delete the report again
+        generate_report_task,
+        # TODO: ensure that any images that need resizing are done so before this is attempted
+        register_submission_task,
+        obtain_submission_reference_task,
```

Comment:
I think `register_submission` (or rather the plugin being called) handles everything that has to do with sending stuff to the registration backend.

If this includes documents (attachments), it can take a while, even though the Zaak is probably already registered and the zaaknumber is technically known. Because uploading documents takes longer, an internal reference is generated which is not really needed. Maybe we should split up the "registering the submission" and "uploading the documents".

But, this sounds like another ticket...

Reply:
Separate ticket indeed. Possibly we can use a separate celery task that is `.delay(...)`-ed as part of the plugin registration so that we get the minimum required "result" asap and can continue. But for the e-mail backend for example, we _do_ need the resized attachments before sending the e-mail, since they are added as attachment to the e-mail.